src/commands.cpp

```diff
@@ -1204,7 +1211,88 @@ bool CommandPutNodes::procresult(Result r)
     }
 #endif
 
-    if (r.wasErrorOrOK())
+    if (r.hasJsonArray() || r.hasJsonObject())
+    {
+        client->sendkeyrewrites();
+
+        // The response is a sparse array indicating the nodes that failed, and the corresponding error code.
+        // If the first three nodes failed, the response would be e.g. [-9,-9,-9].  Success is []
+        // If the second and third node failed, the response would change to {"1":-9,"2":-9}.
+
+        unsigned arrayIndex = 0;
+        for (;;)
+        {
+            if (r.hasJsonArray())
+            {
+                if (!client->json.isnumeric())
+                {
+                    break;
+                }
+                assert(arrayIndex < nn.size());
+                if (arrayIndex < nn.size())
+                {
+                    nn[arrayIndex].mError = error(client->json.getint());
+                }
+            }
+            else
+            {
+                string index, errorCode;
+                if (client->json.storeobject(&index) && !client->json.storeobject(&errorCode))
+                {
+                    arrayIndex = unsigned(atoi(index.c_str()));
+                    nn[arrayIndex].mError = error(atoi(errorCode.c_str()));
+                }
+                else break;
+            }
+        }
+
+#ifdef DEBUG
+        for (auto& n : nn)
+        {
+            // double check we got a node, or know the error why it didn't get created
+            if (!((n.added && n.mAddedHandle != UNDEF && !n.mError) ||
+                 (!n.added && n.mAddedHandle == UNDEF && n.mError)))
+            {
+                assert(false);
+            }
+        }
+#endif
+
+#ifdef ENABLE_SYNC
+        if (source == PUTNODES_SYNC)
+        {
+            client->app->putnodes_result(API_OK, type, nn);
+            client->putnodes_sync_result(API_OK, nn);
+        }
+        else
+#endif
+        if (source == PUTNODES_APP)
+        {
+//#ifdef ENABLE_SYNC        // should be done by actionpacket code
+//            if (!ISUNDEF(targethandle))
+//            {
+//                Node *parent = client->nodebyhandle(targethandle);
+//                if (parent && parent->localnode)
+//                {
+//                    // A node has been added by a regular (non sync) putnodes
+//                    // inside a synced folder, so force a syncdown to detect
+//                    // and sync the changes.
+//                    client->syncdownrequired = true;
+//                }
+//            }
+//#endif
+            assert(!emptyResponse);  // todo: can this emptyResponse case occur now that we use actionpackets?
```

Comment:
Yes, I think it's better to keep it, the old code reported ENOENT in this case, we will continue to do so but by checking the actionpacket data instead of the command response.  Updated